### ISSUE TITLE
Unable to train a provided example (bad results after 55M of samples).

### ISSUE THREAD

**Issue by AGPX** [xbpeng/DeepMimic#159]
Hello @xbpeng,

running the original unmodified code, I get the following result after 55 million samples for 'train_amp_strike_humanoid3d_walk_punch_args.txt':

https://youtu.be/LZmtoZxfgV4

as you can see, the actor is still unable to walk properly. It's normal? How many iterations are needed to get the pre-trained result?

Consider that I stopped training (CTRL + C) and restarted after adding the line: '--model_files output/agent0_model.ckpt' in order to reload the model trained so far (avoiding starting from scratch). Could this be the cause? Is it supported to stop and continue the training at a later time?

Thanks in advance,

G.

**Comment by xbpeng** [comment]
55 million samples is not a lot, especially for a challenging task like this one. You will need to train this policy with about 300-500 million samples.

The code doesn't currently support resuming training. But I think what you did, with reloading a previous checkpoint, should be ok. But I'm not sure if there are some internal states that may not be initialized properly.

**Comment by AGPX** [comment]
> The code doesn't currently support resuming training. I think what you did, with reloading a previous checkpoint, should be ok. But I'm not sure if there are some internal states that may not be initialized properly.

Doing such a long training in a single, continuous, run is quite prohibitive for me: the ability to restart from a checkpoint is crucial for me. I'm going to test if reloading the checkpoint gives me good results (please let me know if you think of any states that may not be initialized properly!). To make the thing easier, I have update the following code:

From rl_world.py, line 84: 

```
if curr_model_file != 'none':
    curr_agent.load_model(curr_model_file)
```

to

```
if curr_model_file != 'none':
    if os.path.isfile(curr_model_file + ".index"):
        curr_agent.load_model(curr_model_file)
```

in order to avoid errors when the model doesn't exists yet (i.e. for the first run).

Anyway, 300-500 million of samples is a very HUGE number. With these numbers, it takes a long time to find out that, for example, your reward function is completely wrong! In the AMP strike example (if I had made it myself), I would have concluded that there was some problem in my code, instead it was just a matter of samples!
Do you have any suggestions in this regard, on how to evaluate faster if we have made mistakes?
Speeding up training would be very helpful. Can, using algorithms other than PPO (such as TRPO) or other optimizers (Adam instead of SGD with momentum), reduce such number of samples? Have you done any experiments?

**Comment by xbpeng** [comment]
The code can definitely still be optimized to be faster, but off the top of my head, I don't know if there are any easy changes that will lead to significant speedups. Using TRPO instead of PPO likely won't lead to much speed improvements. You could try using Adam, but I tend to find that SGD leads to better results, even if it might take more samples.

**Comment by AGPX** [comment]
> You could try using Adam, but I tend to find that SGD leads to better results, even if it might take more samples.

Looks like this is not really true: https://medium.com/geekculture/a-2021-guide-to-improving-cnns-optimizers-adam-vs-sgd-495848ac6008

"The main finding of this paper is that by tuning all available hyperparameters at scales in deep learning, more general optimizers never underperform their special cases. In particular, they observe that RMSProp, Adam, and NAdam never underperformed SGD, NESTEROV, or Momentum."

The problem is that people generally use Adam with the default hyperparameters and this can lead to worse generalization performance, but the real problem is that Adam's hyperparameters need to be tuned correctly as well.
